src/UndockedRegFreeWinRT/UndockedRegFreeWinRT/catalog.cpp

```diff
@@ -81,72 +82,373 @@ struct component
+            }
+        }
+    }
     return S_OK;
 }
 
+HRESULT ParseAssemblyIdentityTag(ComPtr<IXmlReader> xmlReader, bool searchDependentAssembly)
+{
+    HRESULT hr = S_OK;
+    LPCWSTR dependentAssemblyFileName = nullptr;
+    hr = xmlReader->MoveToAttributeByName(L"name", nullptr);
+    if (hr != S_OK)
+    {
+        return HRESULT_FROM_WIN32(ERROR_SXS_MANIFEST_PARSE_ERROR);
+    }
+    RETURN_IF_FAILED(xmlReader->GetValue(&dependentAssemblyFileName, nullptr));
+    if (dependentAssemblyFileName == nullptr || !dependentAssemblyFileName[0])
+    {
+        return HRESULT_FROM_WIN32(ERROR_SXS_MANIFEST_PARSE_ERROR);
+    }
+    // Look out for assemblies that have already been parsed so we don't have circular dependencies
+    auto itr = g_assemblies.find(dependentAssemblyFileName);
+    if (itr != g_assemblies.end())
+    {
+        return S_OK; // We've already look through this assembly, no need to look insert or search for it again. 
+    }
+    g_assemblies.insert(dependentAssemblyFileName);;
+    if (searchDependentAssembly)
+    {
+        // Don't fail if we don't find the assembly, in the realworld, manifests may have 
+        // dependentAssembly ref to WinSxS assemblies which this code won't find.
+        SearchAssembly(dependentAssemblyFileName);
+    }
+    return hr;
+}
+
+
+HRESULT SearchAssembly(std::wstring dependentAssemblyFileName)
+{
+    // Following the pattern in https://docs.microsoft.com/en-us/windows/win32/sbscs/assembly-searching-sequence
+    // RegFree also does not support looking in winSxS folder
+    HRESULT hr = S_OK;
+    wchar_t buffer[2048];
+    unsigned long cchBuffer = ARRAYSIZE(buffer);
+    unsigned long cLanguages;
+    GetUserPreferredUILanguages(MUI_LANGUAGE_NAME, &cLanguages, buffer, &cchBuffer);
+    std::wstring lang = buffer;
+    // Search for the preferred UI language
+    hr = SearchAssemblyWithLang(dependentAssemblyFileName, lang);
+    if (hr == S_OK)
+    {
+        return hr;
+    }
+    if (hr == COR_E_FILENOTFOUND || hr == HRESULT_FROM_WIN32(ERROR_PATH_NOT_FOUND))
+    {
+        hr = SearchAssemblyWithLang(dependentAssemblyFileName, lang.substr(0, 2));
+        if (hr == S_OK)
+        {
+            return hr;
+        }
+    }
+    // This wasn't clear in the docs but it seems to fallback to en-US first
+    if (lang != L"en-US")
+    {
+        hr = SearchAssemblyWithLang(dependentAssemblyFileName, L"en-US");
+        if (hr == S_OK)
+        {
+            return hr;
+        }
+        if (hr == COR_E_FILENOTFOUND || hr == HRESULT_FROM_WIN32(ERROR_PATH_NOT_FOUND))
+        {
+            hr = SearchAssemblyWithLang(dependentAssemblyFileName, L"en");
+            if (hr == S_OK)
+            {
+                return hr;
+            }
+        }
+    }
+    // Finally search for a no language version
+    return SearchAssemblyWithLang(dependentAssemblyFileName, L"");
+}
+
+HRESULT SearchAssemblyWithLang(std::wstring dependentAssemblyFileName, std::wstring lang)
```

Comment:
`std::wstring const&` - the caller already has a `std::wstring` and passing by-value induces an unnecessary copy.